File: python/dnn_control/odesystem.py
class ODESystem:
    '''
    This class represents the dynamics in a rotating reference frame (which we have around an asteroid) using ordinary
    differential equations. Implementation is inspired from "Control of Hovering Spacecraft Using Altimetry" by
    S. Sawai et. al.
    '''

    def __init__(self, asteroid):
        self.asteroid = asteroid
        self.thrust = [0.0, 0.0, 0.0]
        self.coef_earth_acceleration_mul_specific_impulse = 0.0
        self.perturbations_acceleration = [0.0, 0.0, 0.0]

    # This function gets called by the scipy ode stepper to integrate the system.
    # The function computes d/dt x = f(x, t). Or in our case: "d_state_dt" = this(state, time)
    def dynamics(self, time, state):
        from math import sqrt

        # implements eq(1) of "Control of Hovering Spacecraft Using Altimetry" by S. Sawai et. al.
        # r'' + 2w x r' + w' x r + w x (w x r) = Fc + Fg,
        # whereas Fc = control acceleration and Fg = gravity acceleration

        d_state_dt = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

        position = [state[0], state[1], state[2]]
        velocity = [state[3], state[4], state[5]]
        mass = state[6]

        # 1/m
        coef_mass = 1.0 / mass

        # g
        gravity_acceleration = self.asteroid.gravity_at_position(position)

        # w, w'
        angular_velocity, angular_acceleration = self.asteroid.angular_velocity_and_acceleration_at_time(time)

        # Fg, Fc
        thrust_acceleration = [0.0, 0.0, 0.0]
        for i in [0, 1, 2]:
            gravity_acceleration[i] *= coef_mass
            thrust_acceleration[i] = self.thrust[i] * coef_mass

        d_state_dt[0] = state[3]
        d_state_dt[1] = state[4]
        d_state_dt[2] = state[5]

        d_state_dt[3] = self.perturbations_acceleration[0] + gravity_acceleration[0] + thrust_acceleration[0] \
                        - 2.0 * angular_velocity[1] * velocity[2] + 2.0 * angular_velocity[2] * velocity[1] \
                        - angular_acceleration[1] * position[2] + angular_acceleration[2] * position[1]\
                        - angular_velocity[0] * angular_velocity[1] * position[1] \
                        + angular_velocity[1] * angular_velocity[1] * position[0] \
                        + angular_velocity[2] * angular_velocity[2] * position[0]\
                        - angular_velocity[0] * angular_velocity[2] * position[2]

        d_state_dt[4] = self.perturbations_acceleration[1] + gravity_acceleration[1] + thrust_acceleration[1] \
                        - 2.0 * angular_velocity[2] * velocity[0] + 2.0 * angular_velocity[0] * velocity[2] \
                        - angular_acceleration[2] * position[0] + angular_acceleration[0] * position[2] \
                        - angular_velocity[1] * angular_velocity[2] * position[2] \
                        + angular_velocity[2] * angular_velocity[2] * position[1] \
                        + angular_velocity[0] * angular_velocity[0] * position[1]\
                        - angular_velocity[0] * angular_velocity[1] * position[0]

        d_state_dt[5] = self.perturbations_acceleration[2] + gravity_acceleration[2] + thrust_acceleration[2] \
                        - 2.0 * angular_velocity[0] * velocity[1] + 2.0 * angular_velocity[1] * velocity[0] \
                        - angular_acceleration[0] * position[1] + angular_acceleration[1] * position[0] \
                        - angular_velocity[0] * angular_velocity[2] * position[0] \
                        + angular_velocity[0] * angular_velocity[0] * position[2] \
                        + angular_velocity[1] * angular_velocity[1] * position[2] \
                        - angular_velocity[1] * angular_velocity[2] * position[1]

        d_state_dt[6] = -sqrt(self.thrust[0] * self.thrust[0]
                             + self.thrust[1] * self.thrust[1]
                             + self.thrust[2] * self.thrust[2]) * self.coef_earth_acceleration_mul_specific_impulse

        return d_state_dt

File: cpp/dnn_control/odesystem.cpp
#include "odesystem.h"
#include "constants.h"
#include "configuration.h"

ODESystem::ODESystem(const Asteroid &asteroid, const Vector3D &perturbations_acceleration, const Vector3D &thrust, const double &spacecraft_specific_impulse, const double &spacecraft_minimum_mass, const double &engine_noise)
    : asteroid_(asteroid) {
    perturbations_acceleration_ = perturbations_acceleration;
    thrust_ = thrust;
    spacecraft_specific_impulse_ = spacecraft_specific_impulse;
    spacecraft_minimum_mass_ = spacecraft_minimum_mass;
    engine_noise_ = engine_noise;
}

ODESystem::ODESystem(const ODESystem &other)
    : asteroid_(other.asteroid_) {
    perturbations_acceleration_ = other.perturbations_acceleration_;
    thrust_ = other.thrust_;
    spacecraft_specific_impulse_ = other.spacecraft_specific_impulse_;
    spacecraft_minimum_mass_ = other.spacecraft_minimum_mass_;
    engine_noise_ = other.engine_noise_;
}

ODESystem::~ODESystem() {

}

void ODESystem::operator ()(const SystemState &state, SystemState &d_state_dt, const double &time) {
    const double mass = state[6];
    // check if spacecraft is out of fuel
    if (mass <= spacecraft_minimum_mass_) {
        throw OutOfFuelException();
    }

    // 1/m
    const double coef_mass = 1.0 / mass;

    Vector3D position;
    Vector3D velocity;
    for (unsigned int i = 0; i < 3; ++i) {
        position[i] = state[i];
        velocity[i] = state[3+i];
    }

    // Fg
    const Vector3D gravity_acceleration = asteroid_.GravityAccelerationAtPosition(position);

    // w, w'
    const boost::tuple<Vector3D, Vector3D> result = asteroid_.AngularVelocityAndAccelerationAtTime(time);
    const Vector3D angular_velocity = boost::get<0>(result);
    const Vector3D angular_acceleration = boost::get<1>(result);

    // Fc
    Vector3D thrust_acceleration;
    for (unsigned int i = 0; i < 3; ++i) {
        thrust_acceleration[i] = thrust_[i] * coef_mass;
    }

    // w' x r
    const Vector3D euler_acceleration = VectorCrossProduct(angular_acceleration, position);

    // w x (w x r)
    const Vector3D centrifugal_acceleration = VectorCrossProduct(angular_velocity, VectorCrossProduct(angular_velocity, position));

    Vector3D tmp;
    for(unsigned int i = 0; i < 3; ++i) {
        tmp[i] = angular_velocity[i] * 2.0;
    }

    // 2w x r'
    const Vector3D coriolis_acceleration = VectorCrossProduct(tmp, velocity);

    for (unsigned int i = 0; i < 3 ;++i) {
        d_state_dt[i] = state[3+i];
        d_state_dt[3+i] = perturbations_acceleration_[i] + gravity_acceleration[i] + thrust_acceleration[i]
                - coriolis_acceleration[i] - euler_acceleration[i] - centrifugal_acceleration[i];
    }

#if ODES_FUEL_ENABLED
    d_state_dt[6] = -sqrt(thrust_[0] * thrust_[0] + thrust_[1] * thrust_[1] + thrust_[2] * thrust_[2]) / ((spacecraft_specific_impulse_ + spacecraft_specific_impulse_ * engine_noise_) * kEarthAcceleration);
#else
    d_state_dt[6] = 0.0;
#endif
}

